### DISCUSSION TITLE
How do I make array of values out of the enum generated by Prisma to inject in zod enums?

### DISCUSSION THREAD

**Discussion by sbahir** [prisma/prisma#15936] [Q&A] [answered]
Hello everyone,

I want to inject Prisma enums in Zod 

Let's assume the current Prisma `enum` 

```
enum Role {
  user
  admin
}
```

and I want to extract the list `['user', 'admin']` to inject it into `z.enum` 

```
import { z } from 'zod';
import { Role } from '@prisma/client';
const roles = Object.keys(Role);

z.enum(roles);
```

I have the following error 
```
No overload matches this call.
  Overload 1 of 2, '(values: readonly [string, ...string[]], params?: RawCreateParams): ZodEnum<[string, ...string[]]>', gave the following error.
```

Did anyone encounter this situation?

Thank you for your help! 

PS: I found [this discussion](https://github.com/prisma/prisma/discussions/10617) but it didn't provide any answer either

**Comment by austincrim** [comment, marked as answer]
Hey @sbahir 👋,

Could you describe your use case for using `z.enum` a little more? That will help with any proposed solution.

Also, does [`z.nativeEnum`](https://github.com/colinhacks/zod#native-enums) fit your use case better?

**Reply by sbahir** [reply]
@austincrim Thank you for your answer! 

I need to extract the enums defined in Prisma dynamically so I can generate the zod schemas accordingly

in this case, I would use it to parse the body of an API query for example

```
export const createUser = z.object({
  body: z.object({
    role: z.enum(roles),
})
})
```

The role here is just an example. I have multiple enums with a high number of options to maintain

Right now, I am maintaining the enum keys both in Prisma and in Zod separately. So it can cause issue in the future if both don't have the same source of truth

I am not sure if that answers your questionj

I tried to use `z.nativeEnum` but it didn't work

Please let me know if you need more details!

**Reply by pantharshit00** [reply]
Hello @sbahir 

I tried to use `z.nativeEnum` instead of `z.enum` in your example here and it works:
```ts
import { PrismaClient, Role } from '@prisma/client'
import { z } from 'zod'


export const createUser = z.object({
  body: z.object({
    role: z.nativeEnum(Role),
  })
})

const result = createUser.parse({
  body:{
    role: "ser"
  }
})

console.log(result)
```

```sh
yarn run v1.22.17
$ ts-node main.ts
/Users/harshitpant/code/reproductions/issue-15936/node_modules/zod/lib/types.js:29
        const error = new ZodError_1.ZodError(ctx.common.issues);
                      ^
ZodError: [
  {
    "received": "ser",
    "code": "invalid_enum_value",
    "options": [
      "user",
      "admin"
    ],
    "path": [
      "body",
      "role"
    ],
    "message": "Invalid enum value. Expected 'user' | 'admin', received 'ser'"
  }
]
    at handleResult (/Users/harshitpant/code/reproductions/issue-15936/node_modules/zod/lib/types.js:29:23)
    at ZodObject.safeParse (/Users/harshitpant/code/reproductions/issue-15936/node_modules/zod/lib/types.js:140:16)
    at ZodObject.parse (/Users/harshitpant/code/reproductions/issue-15936/node_modules/zod/lib/types.js:120:29)
    at Object.<anonymous> (/Users/harshitpant/code/reproductions/issue-15936/main.ts:11:27)
    at Module._compile (node:internal/modules/cjs/loader:1159:14)
    at Module.m._compile (/Users/harshitpant/code/reproductions/issue-15936/node_modules/ts-node/src/index.ts:1618:23)
    at Module._extensions..js (node:internal/modules/cjs/loader:1213:10)
    at Object.require.extensions.<computed> [as .ts] (/Users/harshitpant/code/reproductions/issue-15936/node_modules/ts-node/src/index.ts:1621:12)
    at Module.load (node:internal/modules/cjs/loader:1037:32)
    at Function.Module._load (node:internal/modules/cjs/loader:878:12) {
  issues: [
    {
      received: 'ser',
      code: 'invalid_enum_value',
      options: [Array],
      path: [Array],
      message: "Invalid enum value. Expected 'user' | 'admin', received 'ser'"
    }
  ],
  addIssue: [Function (anonymous)],
  addIssues: [Function (anonymous)],
  errors: [
    {
      received: 'ser',
      code: 'invalid_enum_value',
      options: [Array],
      path: [Array],
      message: "Invalid enum value. Expected 'user' | 'admin', received 'ser'"
    }
  ]
}
error Command failed with exit code 1.
info Visit https://yarnpkg.com/en/docs/cli/run for documentation about this command.
```

With correct value:
```sh
yarn run v1.22.17
$ ts-node main.ts
{ body: { role: 'user' } }
✨  Done in 0.82s.
```

Let me know if you were referring to something else.

**Reply by saadbahir** [reply]
@pantharshit00 thank you very much!! that works !

**Comment by ludralph** [comment]
Hi there,

To keep our discussions organized and focused on the most relevant topics, we’re reviewing and tidying up our backlog. As part of this process, we’re closing discussions that have already been marked as answered but remain open.

If this discussion still requires further input or clarification, feel free to reopen it or start a new one with updated details. Your contributions are invaluable to the community, and we’re here to help!

For more details about our priorities and vision for the future of Prisma ORM, check out our latest blog post: https://www.prisma.io/blog/prisma-orm-manifesto.

Thank you for your understanding and ongoing support of the Prisma community!
